src/Components/Components/test/Lifetime/ComponentApplicationLifetimeTest.cs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Lifetime;
+using Microsoft.AspNetCore.Components.RenderTree;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging.Abstractions;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Components
+{
+    public class ComponentApplicationLifetimeTest
+    {
+        [Fact]
+        public async Task RestoreStateAsync_InitializesStateWithDataFromTheProvidedStore()
+        {
+            // Arrange
+            byte[] data = new byte[] { 0, 1, 2, 3, 4 };
+            var state = new Dictionary<string, byte[]>
+            {
+                ["MyState"] = data
+            };
+            var store = new TestStore(state);
+            var lifetime = new ComponentApplicationLifetime();
+
+            // Act
+            await lifetime.RestoreStateAsync(store);
+
+            // Assert
+            Assert.True(lifetime.State.TryRetrievePersistedState("MyState", out var retrieved));
+            Assert.Empty(state);
+            Assert.Equal(data, retrieved);
+        }
+
+        [Fact]
+        public async Task RestoreStateAsync_ThrowsOnDoubleInitialization()
+        {
+            // Arrange
+            var state = new Dictionary<string, byte[]>
+            {
+                ["MyState"] = new byte[] { 0, 1, 2, 3, 4 }
+            };
+            var store = new TestStore(state);
+            var lifetime = new ComponentApplicationLifetime();
+
+            await lifetime.RestoreStateAsync(store);
+
+            // Assert
+            await Assert.ThrowsAsync<InvalidOperationException>(() => lifetime.RestoreStateAsync(store));
+        }
+
+        [Fact]
+        public async Task PersistStateAsync_SavesPersistedStateToTheStore()
+        {
+            // Arrange
+            var state = new Dictionary<string, byte[]>();
+            var store = new TestStore(state);
+            var lifetime = new ComponentApplicationLifetime();
+
+            var renderer = new TestRenderer();
+            var data = new byte[] { 1, 2, 3, 4 };
+
+            lifetime.State.PersistState("MyState", new byte[] { 1, 2, 3, 4 });
+
+            // Act
+            await lifetime.PersistStateAsync(store, renderer);
+
+            // Assert
+            Assert.True(store.State.TryGetValue("MyState", out var persisted));
+            Assert.Equal(data, persisted);
+        }
+
+        [Fact]
+        public async Task PersistStateAsync_InvokesPauseCallbacksDuringPersist()
```

Comment:
I'm not sure about this, but is it worth adding a test describing something about the error handling behavior? If one of the callbacks throws, do the subsequent ones run? Are there are any guarantees about ordering?

Maybe you don't want to define the order, which is fine. And maybe the error handling rule is "if one of them throws, all subsequent behavior is undefined", which is fine *if* we have no goals to make this scenario recoverable in any way (e.g., persisting as much state as we can).

Reply:
This is a good point. I think it can be worth treating this as "best effort". State is private to each component and a component/service must be able to function without their state (state is just a perf optimization)